Replace VectorStoreGenericDataModel with Dictionary<string, object?>

roji opened the pull request:
Closes #10802
Closes #10468

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -312,26 +317,60 @@ private void TranslateUnary(UnaryExpression unary)
     }
 
-    private bool TryGetField(Expression expression, [NotNullWhen(true)] out string? field)
+    private bool TryBindProperty(Expression expression, [NotNullWhen(true)] out VectorStoreRecordPropertyModel? property)
```
This is duplicated across the connectors - we'll refactor this out to a common utility later.

adamsitnik commented on dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantVectorStoreRecordCollection.cs:
```diff
@@ -235,103 +236,120 @@ public virtual Task DeleteCollectionAsync(CancellationToken cancellationToken =
+                break;
+
+            case IEnumerable<object> k:
+                var objectKeys = k.ToList();
+                keyList = objectKeys[0] switch
+                {
+                    ulong id => objectKeys.Cast<ulong>().ToList(),
+                    Guid id => objectKeys.Cast<Guid>().ToList(),
```
The `ToList` is being called twice for `IEnumerable<object>`, moreover we risk `IndexOutOfRangeException` by accessing `[0]`.

roji replied:
You're right, I was sloppy here. Will reimplement better.

adamsitnik commented on dotnet/src/Connectors/Connectors.Memory.Redis/RedisJsonDynamicDataModelMapper.cs:
```diff
@@ -59,28 +60,23 @@ public VectorStoreGenericDataModel<string> MapFromStorageToDataModel((string Key
 
-            // Replicate null if the property exists but is null.
-            if (sourceValue is null)
+            // Only map properties across that actually exist in the input.
```
Is this consistent with the other mappers? It seemed like we were now about to always express nulls in explicit way (by adding a null value to the dictionary)

roji replied:
We're not very consistent around this, no... I already have a note to discuss this with you guys at some point and arrive at a desired behavior...

In the meantime, this PR should generally not change the behavior - and I think that's the case no? If the property is missing, we skip it (don't populate null), otherwise we populate.

adamsitnik reviewed (commented):
So far I've reviewed  64 out of 174 files and found only nits.

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLVectorStoreRecordCollection.cs:
```diff
@@ -677,15 +590,29 @@ private IVectorStoreRecordMapper<TRecord, JsonObject> InitializeMapper(JsonSeria
     }
 #pragma warning restore CS0618
 
+    private static IEnumerable<AzureCosmosDBNoSQLCompositeKey> GetCompositeKeys(IEnumerable<TKey> keys)
+        => keys switch
+        {
+            IEnumerable<AzureCosmosDBNoSQLCompositeKey> k => k,
+            IEnumerable<string> k => k.Select(key => new AzureCosmosDBNoSQLCompositeKey(recordKey: key, partitionKey: key)),
+            IEnumerable<object> k => k.Select(key => key switch
+            {
+                string s => new AzureCosmosDBNoSQLCompositeKey(recordKey: s, partitionKey: s),
+                AzureCosmosDBNoSQLCompositeKey ck => ck,
+                _ => throw new ArgumentException($"Invalid key type '{key.GetType().Name}'.")
+            }),
+            _ => throw new UnreachableException()
+        };
+
```
Yeah, the idea here was specifically to avoid needless enumeration if the user already provides the right type (`IEnumerable<AzureCosmosDBNoSQLCompositeKey>`), for performance... It may be over-optimizing...

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs:
```diff
@@ -24,10 +24,21 @@ namespace Microsoft.SemanticKernel.Connectors.AzureAISearch;
-public class AzureAISearchVectorStoreRecordCollection<TRecord> :
-    IVectorStoreRecordCollection<string, TRecord>,
+public class AzureAISearchVectorStoreRecordCollection<TRecord>(SearchIndexClient searchIndexClient, string collectionName, AzureAISearchVectorStoreRecordCollectionOptions<TRecord>? options = default)
+    : AzureAISearchVectorStoreRecordCollection<string, TRecord>(searchIndexClient, collectionName, options);
```
Mainly user-facing sugar, nothing more... For cases where there's really just one key type (string), and users are doing POCO mapping (the presumed majority), it just saved them from having to think about the key type altogether.

westey-m replied:
This is going to clash with the PR where we are making all the collections sealed.
FWIW, I'm OK just having the new one that requires a key if it comes down to it.

adamsitnik replied:
> Mainly user-facing sugar, nothing more

FWIW I don't believe it's worth it then, but I don't have a strong opinion about it.

roji replied:
OK, yeah, if we're making everything sealed, I think it makes sense to back this out then.

roji commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStore.cs:
```diff
@@ -52,12 +52,12 @@ public virtual IVectorStoreRecordCollection<TKey, TRecord> GetCollection<TKey, T
+        if (typeof(TKey) != typeof(string) && typeof(TKey) != typeof(object))
         {
-            throw new NotSupportedException("Only string keys are supported.");
+            throw new NotSupportedException("Only string keys are supported (and object for dynamic mapping)");
```
Yes - this PR specifically removs all mention of "generic" in this context (or tries to), replacing it with dynamic. It's true that C# has a (mostly deprecated) dynamic language feature, but I see little chance of actual confusion, plus the actual intent/scenario here in MEVD is actually really close to what C# dynamic tried to achieve; we're just not using the `dynamic` type to do it (BTW we actually could - it might have been a pretty good match - if only it were were support in LINQ expression trees).

Let me know if you think the term dynamic (and/or dynamic mapping) is problematic.

westey-m replied:
Dynamic seems reasonable to me. I agree that it's more or less matching the intended use case, and this is the only dynamic access method we are going to support.

westey-m commented on dotnet/samples/GettingStartedWithVectorStores/Step5_Use_DynamicDataModel.cs:
```diff
@@ -8,13 +8,13 @@
 /// <summary>
-/// Example that shows that you can use the generic data model to interact with a vector database.
-/// This makes it possible to use the vector store abstractions without having to create your own data model.
+/// Example that shows that you can use the dynamic data modling to interact with a vector database.
```
```suggestion
/// Example that shows that you can use the dynamic data modeling to interact with a vector database.
```

westey-m commented on dotnet/samples/Concepts/Memory/VectorStore_DynamicDataModel_Interop.cs:
```diff
@@ -12,15 +12,15 @@
+/// 2. Upsert data using a custom data model and retrieve it from the vector store using the dynamic data modeling.
 /// </summary>
-public class VectorStore_GenericDataModel_Interop(ITestOutputHelper output, VectorStoreQdrantContainerFixture qdrantFixture) : BaseTest(output), IClassFixture<VectorStoreQdrantContainerFixture>
+public class VectorStore_DynamicDataModel_Interop(ITestOutputHelper output, VectorStoreQdrantContainerFixture qdrantFixture) : BaseTest(output), IClassFixture<VectorStoreQdrantContainerFixture>
```
Most samples are referenced from a README.md in the samples folder, so please also check if the file reference to this needs to be renamed there.

westey-m commented on dotnet/src/Connectors/Connectors.AzureCosmosDBNoSQL.UnitTests/AzureCosmosDBNoSQLVectorStoreRecordCollectionTests.cs:
```diff
@@ -243,18 +243,22 @@ public async Task DeleteInvokesValidMethodsAsync(
         }
         else
         {
+            var sut = new AzureCosmosDBNoSQLVectorStoreRecordCollection<string, AzureCosmosDBNoSQLHotel>(
+    this._mockDatabase.Object,
+    "collection");
```
```suggestion
            var sut = new AzureCosmosDBNoSQLVectorStoreRecordCollection<string, AzureCosmosDBNoSQLHotel>(
                this._mockDatabase.Object,
                "collection");
```

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchFilterTranslator.cs:
```diff
@@ -312,26 +317,60 @@ private void TranslateUnary(UnaryExpression unary)
+
+        if (convertedClrType is not null && convertedClrType != property.Type)
+        {
+            throw new InvalidCastException();
```
Would be useful to give some more information here, e.g. the name of the property that is being compared.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchVectorStoreRecordCollection.cs:
```diff
@@ -693,4 +709,20 @@ private static ReadOnlyMemory<float> VerifyVectorParam<TVector>(TVector vector)
+            _ => throw new UnreachableException("string key should have been validated during model building")
+        };
+
+        Verify.NotNullOrWhiteSpace(stringKey);
```
This throws an ArgumentException using the name of the variable used here `stringKey`, which wouldn't make sense for callers since there is no param with that name on their signature.  Consider passing the second param to NotNullOrWhiteSpace and explicitly naming the param that they would see.

westey-m replied:
Same for similar methods in other connectors too.

roji replied:
Good catch.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLVectorStoreRecordCollection.cs:
```diff
@@ -543,88 +538,6 @@ private static VectorDataType GetDataType(Type vectorDataType, string vectorProp
-        }
-    }
-
-    private async Task<AzureCosmosDBNoSQLCompositeKey> InternalUpsertAsync(
```
Can we please not do this kind of refactoring with other changes.  It makes it really hard to review the main changes.

roji replied:
I promise I really did my best to prevent unrelated refactorings to slip in... But in this particular case, I did need to merge the two versions of each method - we previously had one for string and one for AzureCosmosDBNoSQLCompositeKey, whereas now we only have one that deals with both (since we're generic over TKey)... Hopefully that makes sense.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.InMemory/InMemoryVectorStore.cs:
```diff
@@ -47,10 +47,10 @@ public IVectorStoreRecordCollection<TKey, TRecord> GetCollection<TKey, TRecord>(
-            this._internalCollectionTypes,
-            name,
-            new() { VectorStoreRecordDefinition = vectorStoreRecordDefinition }) as IVectorStoreRecordCollection<TKey, TRecord>;
+             this._internalCollection,
```
NIT: Looks like this is now 5 spaces

roji replied:
Good catch - am surprised dotnet format didn't error on this...

westey-m commented on dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantVectorStore.cs:
```diff
@@ -61,12 +61,12 @@ public virtual IVectorStoreRecordCollection<TKey, TRecord> GetCollection<TKey, T
 #pragma warning restore CS0618
 
-        if (typeof(TKey) != typeof(ulong) && typeof(TKey) != typeof(Guid))
+        if (typeof(TKey) != typeof(ulong) && typeof(TKey) != typeof(Guid) && typeof(TKey) != typeof(object))
```
Now that we support providing any key type to the collection for most collection types, we should move this validation into the collection constructor, otherwise someone can construct the collection like this:
`new QdrantVectorStoreRecordCollection<string, MyGenericKeyRecord<Guid>>(...)`

We validate the key type of the data model after all, but I don't believe we are validating the key type provided to the collection anywhere else.

roji replied:
Good idea. This also relates to #11355: even if we validate the (generic) key type, we still need to validate that it's actually the same as the type in the model... But I'll do this change for now.

adamsitnik reviewed (approved):
LGTM, thank you for adding a lot of new conformance tests @roji!

Also, it's great to see the filters support the dynamic model.

roji commented on dotnet/src/Connectors/Connectors.Memory.MongoDB/MongoDBVectorStoreRecordCollection.cs:
```diff
@@ -637,5 +656,22 @@ private static Array VerifyVectorParam<TVector>(TVector vector)
         };
     }
+
+    private string GetStringKey(TKey key)
```
The problem is that this is is string-specific (e.g. it checks that it doesn't contain whitespace). Plus most other providers accept multiple key types, so it wouldn't useful for them... I know it's a bit of duplication, but it's small/harmless enough that I'd rather not introduce a full public API for it in MEVD itself (even if it's only connector-facing).

roji commented on dotnet/src/Connectors/Connectors.Memory.Weaviate/WeaviateVectorStoreRecordCollection.cs:
```diff
@@ -21,7 +22,21 @@ namespace Microsoft.SemanticKernel.Connectors.Weaviate;
+/// <typeparam name="TKey">The data type of the record key. Can be either <see cref="Guid"/>, or <see cref="object"/> for dynamic mapping.</typeparam>
+/// <typeparam name="TRecord">The data model to use for adding, updating and retrieving data from storage.</typeparam>
+#pragma warning disable CA1711 // Identifiers should not have incorrect suffix
+public class WeaviateVectorStoreRecordCollection<TKey, TRecord> : IVectorStoreRecordCollection<TKey, TRecord>, IKeywordHybridSearch<TRecord>
```
Update: backing out the non-generic sub-class, as per https://github.com/microsoft/semantic-kernel/pull/11392#discussion_r2031263566.

roji commented on dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantVectorStoreRecordCollection.cs:
```diff
@@ -235,103 +236,120 @@ public virtual Task DeleteCollectionAsync(CancellationToken cancellationToken =
-    }
+        // Create options.
+        var pointsIds = keys
+            .Select(key => key switch
```
Will do.

roji commented on dotnet/src/VectorDataIntegrationTests/InMemoryIntegrationTests/Filter/InMemoryBasicFilterTests.cs:
```diff
@@ -5,13 +5,15 @@
     {
         public override TestStore TestStore => InMemoryTestStore.Instance;
+
+        public override bool TestDynamic => false;
```
Yeah, it's a bit annoying... Adding the following comment to provide context:

> BaseFilterTests attempts to create two InMemoryVectorStoreRecordCollection with different .NET types:
> 1. One for strongly-typed mapping (TRecord=FilterRecord)
> 2. One for dynamic mapping (TRecord=Dictionary<string, object?>)
> Unfortunately, InMemoryVectorStore does not allow mapping the same collection name to different types;
> at the same time, it simply evaluates all filtering via .NET AsQueryable(), so actual test coverage
> isn't very important here. So we disable the dynamic tests.

roji commented on dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordDataPropertyModel.cs:
```diff
@@ -29,47 +28,6 @@ public class VectorStoreRecordDataPropertyModel(string modelName, Type type) : V
-            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(VectorStoreGenericDataModel<>))
-            {
-                var dataProperty = type.GetProperty("Data")!;
-                var dictionary = (Dictionary<string, object?>)dataProperty.GetValue(record)!;
```
Same :)

roji commented on dotnet/src/Connectors/VectorData.Abstractions/ConnectorSupport/VectorStoreRecordPropertyModel.cs:
```diff
@@ -53,44 +54,57 @@ public string StorageName
+        if (this.PropertyInfo is null)
         {
-            // We have a .NET property (non-dynamic POCO mapping)
+            if (record.GetType() == typeof(Dictionary<string, object?>))
+            {
+                var dictionary = (Dictionary<string, object?>)record;
```
Good catch, thanks!

roji commented on dotnet/src/VectorDataIntegrationTests/SqlServerIntegrationTests/SqlServerCommandBuilderTests.cs:
```diff
@@ -138,7 +138,7 @@ PRIMARY KEY ([id])
         }
 
-        AssertEqualIgnoreNewLines(expectedCommand, command.CommandText);
+        Assert.Equal(expectedCommand, command.CommandText, ignoreLineEndingDifferences: true);
```
Yeah there are some other options xUnit provides, check it out..

roji commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Xunit/ConditionalTheoryDiscoverer.cs:
```diff
@@ -0,0 +1,39 @@
+///     Used dynamically from <see cref="ConditionalTheoryAttribute" />.
+///     Make sure to update that class if you move this type.
+/// </summary>
+public class ConditionalTheoryDiscoverer(IMessageSink messageSink) : TheoryDiscoverer(messageSink)
```
Ah right, I did fix it here 😅

roji commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/CRUD/BatchConformanceTests.cs:
```diff
@@ -44,27 +44,20 @@ private async Task GetBatchAsyncReturnsInsertedRecords(bool includeVectors)
     }
 
     [ConditionalFact]
-    public Task UpsertBatchAsyncCanInsertNewRecord_WithVectors()
```
I think it's just that WithVectors affects the get operations only, and that's already tested elsewhere (e.g. GetBatchAsyncReturnsInsertedRecords_WithVectors) - we probably don't need to multiply every test twice for WithVectors.

(When we finish all the critical stuff, I plan to do a pass over all the conformance tests to eliminate some duplicate coverage, refactor and clean up a bit...)

adamsitnik reviewed (approved):
LGTM, thank you @roji !

roji commented on dotnet/src/VectorDataIntegrationTests/VectorDataIntegrationTests/Filter/BasicFilterTests.cs:
```diff
@@ -226,15 +306,13 @@ public virtual Task Legacy_AnyTagEqualTo_List()
         bool expectZeroResults = false,
         bool expectAllResults = false)
     {
-        var expected = this.GetOrderedRecords(fixture.TestData.AsQueryable().Where(filter));
```
@adamsitnik FYI I've removed the ordering that you added here for filteirng GetAsync; this got a bit complicated when merging with the dynamic work in this PR.

I also think that ordering should not be covered by this test suite, but rather in a separate one - tests here really are about different types of filter expressions, and we should try to keep things separate as much as possible.

I'm also not sure if the ordering coverage was actually sufficient here: it's likely that the tests could pass by accident, since the database happens to return records in a certain order. In other words, for full coverage we need separate tests which specifically exercise ordering features (ascending, descending...), and which ensure that the user's specified ordering is actually applied by the database (e.g. sort twice, once in ascending and once in descending, to fully verify things).

I'll take some time to clean up this area of the tests soon.

roji commented:
@westey-m @adamsitnik I addressed all the comments (AFAIK) and rebased on top of the latest branch - please take a look. Most extra modifications you're seeing are just about adding the mandatory TKey parameter everywhere.

westey-m commented on dotnet/src/Connectors/Connectors.Memory.AzureCosmosDBNoSQL/AzureCosmosDBNoSQLServiceCollectionExtensions.cs:
```diff
@@ -103,7 +103,7 @@ public static IServiceCollection AddAzureCosmosDBNoSQLVectorStoreRecordCollectio
                 var selectedOptions = options ?? sp.GetService<AzureCosmosDBNoSQLVectorStoreRecordCollectionOptions<TRecord>>();
 
-                return new AzureCosmosDBNoSQLVectorStoreRecordCollection<TRecord>(database, collectionName, selectedOptions);
+                return new AzureCosmosDBNoSQLVectorStoreRecordCollection<string, TRecord>(database, collectionName, selectedOptions);
```
I think hardcoding the key on line 99 above may be problematic for the case where the user is using Dictionary<string, object>.  Something we can handle in a separate PR, but we probably want to revisit this as part of the DI work.

roji replied:
Yeah, agreed we should think about it.

The current DI methods aren't generic over TKey (where there's only one type, as here). I think it could be reasonable to leave it like this - the 95% usage case is users doing POCO mapping, and IMHO it makes sense to keep things super simple for them. Dynamic mapping scenarios are likely going to be more complex anyway, and so I'm not sure we need to provide the niceness here (i.e. layers like KernelMemory can register whatever they want in DI themselves, without necessarily going through our DI methods).

Options:

1. Expose TKey on the DI methods. This would allow users to use them for both POCO and dynamic mapping, but makes things slightly more cumbersome for the 95% POCO case.
2. Decide on the key type ourselves in the DI method, i.e. use object if TRecord is dictionary, otherwise string. That's a bit magical/non-obvious - users would have to know to get injected with the different IVectorStoreRecordCollection interfaces based on this. I'm not a huge fan.
3. Leave things as they are - if users try to pass a dictionary as TRecord, validation should fail at that point (we may need to do something to make that happen though).

roji replied:
One more thought on this...

Looking a tiny bit into the future, we need to make everything compatible with trimming and NativeAOT. The DI method as it is - accepting an arbitrary user POCO type - will not be trimming compatible, since it internally uses reflection to access the properties; in other words, there's nothing referencing the POCO's properties at compile time, and no way the linker would know that it needs to preserve them. So this method will need to be annotated with `[RequiresUnreferencedCode]` (that's what #10956 tracks).

The only solution to this problem - which also addresses hierarchical data models - is a source generator, which would generate accessors for the properties (recursively); the generated artifact of the source generator would have to be passed in as another argument (just like System.Text.Json accepts a JsonSerializerContext ([docs](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation)) - that's a new DI method overload that we'll need to introduce here; the existing one will need to remain trimming-incompatible.

However, dynamic mapping doesn't have any of these issues - no .NET type is involved (only Dictionary and dictionary lookups), and so it's always compatible with trimming/NativeAOT (a source generator isn't needed, nor is one even possible, as there's no type for it to process). So if we want to allow users to register dynamically-mapped collections in DI, we'll almost certainly want a separate overload, i.e. AddDynamicAzureCosmosDBNoSQLVectorStoreRecordCollection (yes we really need to shorten these names). This overload would **not** have `[RequiresUnreferenceCode]`, would not accept any source-generated artifacts, and would be hard-wired to work with `Dictionary<string, object?>` (and an object key).

In other words, we're in any case going in a direction where different methods are needed for registering a POCO-based collection, and for registering a dynamic collection. So it's probably OK for the current method to just throw if a user attempts to use it with Dictionary.
